File: c++/p301-p400/leetcode306.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author :   Shen                                                         *
*   @name   :   Leetcode 306 Additive Number                                 *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

class Solution
{
public:
    typedef long long int64;
    
    bool isAdditiveNumber(string num)
    {
        if (num.size() < 3) return false;
        return dfs(num, 0);
    }

    bool dfs(string s, int start)
    {
        for (int i = start, i_end = s.size() - 2; i < i_end; i++)
        {
            int64 a = 0, b = 0;
            for (int j = i + 1, j_end = s.size(); j < j_end; j++)
            {
                
                a = std::stoll(s.substr(start, i + 1 - start));
                b = std::stoll(s.substr(i + 1, j - i));
                string ans = std::to_string(a + b);
                string sc = s.substr(j + 1, ans.size());
                if (ans == sc)
                {
                    if (j + 1 + ans.size() == s.size()) return true;
                    if (dfs(s, i + 1)) return true;
                    break;
                }
                if (b == 0) break;
            }
            if (a == 0) return false;
        }
        return false;
    }
};

void test()
{
    Solution sol;
    vector<string> tests = { "112358", "199100199", "111111", "010102" };
    vector<bool> answer = { true, true, false, false };
    for (int i = 0 ; i < 4; i++)
    {
        printf("tests= %s, result= %s, answer= %s\n", tests[i].c_str(), sol.isAdditiveNumber(tests[i])?"true":"false", answer[i]?"true":"false");
    }
}

int main() { test(); return 0; }

File: Archive#03-301-400/306-Additive Number/leetcode306.py
# 1123581321 -> [1,1,2,3,5,18,13] => True
# 100101201302 -> [100,101,201,302] => True
# 112211221122 => False
class Solution:
    def isAdditiveNumber(self, num: str) -> bool:
        if len(num) < 3: return False
        return self.dfs(num, 0)
    
    def dfs(self, s, start):
        # select a, b from s, then check a + b = rest
        # print(">>> dfs(s, ", start, ")")
        for i in range(start, len(s) - 2):
            for j in range(i + 1, len(s)):
                a = int(s[start:i + 1])
                b = int(s[i + 1:j + 1])
                ans = str(a + b)
                s3 = s[j + 1:j + 1 + len(ans)]
                # print("i= ", i, "j= ", j, "a= ", a, "b= ", b, "ans= ", ans, "s3= ", s3, )
                if ans == s3:
                    if j + 1 + len(ans) == len(s): return True
                    rest = self.dfs(s, i + 1)
                    if rest: return True
                    break
                if b == 0: break
            if a == 0: return False
        return False


def main():
    runner = Solution().isAdditiveNumber
    tests = ['112358', '199100199', '111111', '010102']
    answers = [True, True, False, False]
    results = [runner(_) for _ in tests]
    for (_1, _2, _3) in zip(tests, results, answers):
        print("tests= ", _1, "result= ", _2, "answer= ", _3)


if __name__ == '__main__':
    main()
